## DATA CLEANING

**Adriana de Vicente**


**Irma Sánchez**

In this notebook we make some changes on our dataset for the proceed to make the prediction model

# Import packages

In [1]:
import warnings
warnings.filterwarnings('ignore')
from copy import deepcopy
import pickle
import pandas as pd
import numpy as np

# Load business data

In [2]:
df = pd.read_pickle('../data/processed/businessdf_clean.pickle')
df.head()

business_id             business_name           city US_state  \
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ  Santa Barbara       CA   
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store         Affton       MO   
2  tUFrWirKiKi_TAnsVWINQQ                    Target         Tucson       AZ   
3  MTSW4McQd7CbVtyjqoe9mw        St Honore Pastries   Philadelphia       PA   
4  mWMc6_wTdE0EUBKIGXDVfA  Perkiomen Valley Brewery     Green Lane       PA   

    latitude   longitude  average_stars_per_business  \
0  34.426679 -119.711197                         5.0   
1  38.551126  -90.335695                         3.0   
2  32.223236 -110.880452                         3.5   
3  39.955505  -75.155564                         4.0   
4  40.338183  -75.471659                         4.5   

   review_count_per_business  business_status  \
0                          7                0   
1                         15                1   
2                         22                0   
3                         80                1   
4                         13                1   

                                          categories  ...  meal_brunch  \
0  Doctors, Traditional Chinese Medicine, Naturop...  ...          0.0   
1  Shipping Centers, Local Services, Notaries, Ma...  ...          0.0   
2  Department Stores, Shopping, Fashion, Home & G...  ...          0.0   
3  Restaurants, Food, Bubble Tea, Coffee & Tea, B...  ...          0.0   
4                          Brewpubs, Breweries, Food  ...          0.0   

   meal_dessert  meal_dinner  meal_latenight  meal_lunch  parking_garage  \
0           0.0          0.0             0.0         0.0             0.0   
1           0.0          0.0             0.0         0.0             0.0   
2           0.0          0.0             0.0         0.0             0.0   
3           0.0          0.0             0.0         0.0             0.0   
4           0.0          0.0             0.0         0.0             0.0   

   parking_lot  parking_street  parking_valet  parking_validated  
0          0.0             0.0            0.0                0.0  
1          0.0             0.0            0.0                0.0  
2          0.0             0.0            0.0                0.0  
3          0.0             0.0            0.0                0.0  
4          0.0             0.0            0.0                0.0  

[5 rows x 104 columns]

you can read a pickle file and then preview the data in the resulting dataframe, all in a single line of code.

# Extract Restaurant or food store information

In [3]:
# Specify rows data that contains 'Restaurants' or 'Food' in the 
restaurants = df[df['categories'].str.contains('Restaurant|Food') == True]

the restaurants dataframe contains only the rows from the original dataframe where the value in the categories column contains the string "Restaurant" or "Food

In [4]:
# Check the shape
restaurants.shape

(64629, 104)

he output shows that there are 64,629 rows and 104 columns in the restaurants dataframe.

In [5]:
# Check if I extracted restaurant and food data successfully 
restaurants.categories[:5]

3    Restaurants, Food, Bubble Tea, Coffee & Tea, B...
4                            Brewpubs, Breweries, Food
5    Burgers, Fast Food, Sandwiches, Food, Ice Crea...
8    Pubs, Restaurants, Italian, Bars, American (Tr...
9    Ice Cream & Frozen Yogurt, Fast Food, Burgers,...
Name: categories, dtype: object

In [6]:
# Put labels to business information for future use
Business_type = ['Restaurants', 'Food']
for b in Business_type:
    restaurants.loc[restaurants['categories'].str.contains(b),'business_type'] = b

The previous code is used to label businesses in the "restaurants" dataframe as belonging to the "Restaurants" or "Food" category. The code iterates through a list of business types called "Business_type" and uses the "str.contains()" method to check if the "categories" column in the "restaurants" dataframe contains the current business type. If it does, the code adds a new column called "business_type" to the "restaurants" dataframe and assigns the current business type as the value for that row.

In [7]:
# Those with NaN values after the above step are 'Restaurant Supplies', thus, I decided to drop the information
restaurants['business_type'].dropna(inplace=True)

This code is for labeling the businesses in the 'restaurants' dataframe based on the categories they belong to. The labels 'Restaurants' and 'Food' are defined in the list 'Business_type'. The code then uses the 'loc' function to create a new column 'business_type' and assign it the values 'Restaurants' or 'Food' if they are present in the 'categories' column. If a business does not belong to either of these categories, its 'business_type' value will be NaN. The last line of code is then used to drop these NaN values from the 'business_type' column.

# Create a column of couisine information

In [8]:
# Create a cuisine column based on a popular cuisine list
cuisines = ['African', 'Algerian', 'American', 'Belgian', 'Brazilian', 'British', \
            'Cajun', 'Canadian', 'Chinese', 'Cuban', 'Egyptian', 'French', \
            'German', 'Greek', 'Haitian', 'Hawaiian', 'Indian', 'Irish', 'Italian', 
            'Japanese', 'Jewish', 'Kenyan', 'Korean', 'Latvian', \
            'Libyan', 'Mediterranean', 'Mexican', 'Middle Eastern', 'Mormon', \
            'Nigerian', 'Peruvian', 'Polish', 'Portuguese', 'Russian', \
            'Salvadorian', 'Scottish', 'Spanish', 'Swedish', \
            'Tahitian', 'Thai', 'Tibetan', 'Turkish', 'Vietnamese','Welsh'\
            'Venezuelan', 'Argentine', 'Cantonese', 'Salvadoran', 'Arabian',\
            'Moroccan', 'Russian', 'Uzbek', 'Armenian', 'Lebanese', 'Indonesian', \
            'Singaporean']

The above code creates a list of cuisines. This list will be used to categorize the types of cuisines offered by restaurants in the data.

In [9]:
# Find a cuisine name in the 'categories' column and create a new colomn (or fill 1 in the relevant column)
# E.g. When a cell contains 'American', it creates an 'American' column or fill 1 in the 'American' column if it exists already
for c in cuisines:
    restaurants.loc[restaurants['categories'].str.contains(c),'cuisine'] = c

The above code searches for specific cuisine names in the 'categories' column of the 'restaurants' dataframe and creates a new 'cuisine' column or fills a 1 in the relevant column if it already exists

# Create a column of food type information

In [10]:
# Create a food type column based on a popular cuisine list
Food_type = ['Arabian', 'Armenian', 'Asian Fusion', 'Barbeque', 'Basque', 
            'Breakfast & Brunch', 'Buffets', 'Burgers', 'Cafes', \
            'Cafeteria', 'Cajun', 'Cheesesteaks', 'Chicken Shops', \
            'Chicken Wings', 'Coffee & Tea', 'Comfort Food', 'Creperies', \
            'Delis', 'Dim Sum', 'Diners', 'Falafel', 'Fast Food', 'Fish & Chips', \
            'Fondue', 'Food Court', 'Food Stands', 'Gastropubs', 'Halal', \
            'Hot Dogs', 'Hot Pot', 'Kebab', 'Kosher', 'Moroccan', 'Noodles',\
            'Pizza', 'Ramen', 'Russian', 'Salad', 'Salvadoran', 'Sandwiches', \
            'Seafood', 'Soup', 'Steakhouses', 'Sushi Bars', 'Tacos', 'Tapas Bars', \
            'Tapas Bars', 'Uzbek', 'Waffles', 'Wraps', ]

The above code creates a list called 'Food_type' which includes various types of food. It is intended to be used later to create a new column in the 'restaurants' dataframe called 'food_type'. This new column will contain the type of food served at each restaurant. If a restaurant serves a particular type of food, it will fill 1 in the relevant column for that type of food.

In [11]:
# Do the same as 'cuisine' columns
for f in Food_type:
    restaurants.loc[restaurants['categories'].str.contains(f),'food_type'] = f

This code creates a new column in the 'restaurants' dataframe called 'food_type'. It looks for a specific food type in the 'categories' column of the dataframe and fills the corresponding cell in the 'food_type' column with the name of the food type if it is found. The list of food types being searched for is specified in the 'Food_type' list.

# Dummify columns 

In [12]:
dummy_business_type = pd.get_dummies(restaurants[['business_type']], drop_first=True)

In [13]:
dummy_cuisine_and_food = pd.get_dummies(restaurants[['cuisine', 'food_type']])

In [14]:
restaurants_ver2 = pd.concat([restaurants, dummy_business_type, dummy_cuisine_and_food], axis=1)

In [15]:
restaurants_ver2 = restaurants_ver2.drop(['business_type', 'categories', 'cuisine', 'food_type'], axis=1)

This code creates dummy variables for the columns 'business_type', 'cuisine', and 'food_type' in the 'restaurants' dataframe. Dummy variables are numerical representations of categorical variables, where each unique category is represented by a separate binary column. The 'pd.get_dummies()' function is used to create the dummy variables and the 'drop_first=True' argument is used to drop the first level of each categorical variable, so that there is no perfect multicollinearity between the dummy variables. The resulting dummy variables are then concatenated to the original 'restaurants' dataframe using the 'pd.concat()' function, with the 'axis=1' argument indicating that the concatenation should be performed along the columns. Finally, the original categorical columns 'business_type', 'categories', 'cuisine', and 'food_type' are dropped from the dataframe using the 'df.drop()' function.

# Feature generation 

## Create a column for total checkin counts

In [16]:
# Load check-in data
checkins = pd.read_pickle('../data/processed/checkinjson.pickle')
checkins.head()

business_id                                               date
0  ---kPU91CF4Lq2-WlRu9Lw  2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...
1  --0iUa4sNDFiZFrAdIWhZQ  2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...
2  --30_8IhuyMHbSOcNWd6DQ           2013-06-14 23:29:17, 2014-08-13 23:20:22
3  --7PUidqRWpRSpXebiyxTg  2011-02-15 17:12:00, 2011-07-28 02:46:10, 2012...
4  --7jw19RH9JKXgFohspgQw  2014-04-21 20:42:11, 2014-04-28 21:04:46, 2014...

The code loads the data stored in the file 'checkinjson.pickle' and saves it to a variable called 'checkins'. It then displays the first few rows of the data using the head() function.

In [38]:
# Create a new DataFrame to hold the timeseries of the checkins
#df_checkins = pd.DataFrame()

#for id in range(len(checkins.date)):
    
#    df_checkins_id = pd.DataFrame()
    
#    split = checkins.date[id].split(',')
#    df_checkins_id['Checkin_date'] = pd.to_datetime(split)
#    df_checkins_id['Business_id'] = checkins.business_id[id]
#    df_checkins_id['Year'] = df_checkins_id['Checkin_date'].dt.year
    

#    df_checkins_id = df_checkins_id.groupby(['Business_id','Year']).agg(['count']).reset_index()
#    print(df_checkins_id)

#    df_checkins = pd.concat([df_checkins, df_checkins_id])

The code above creates a new empty dataframe called 'df_checkins' to hold the timeseries of the check-ins data. Then, it loops through each row of the 'checkins' dataframe, creates a new dataframe called 'df_checkins_id' for each row, and fills it with the check-in dates and business IDs. The dates are split by the comma separator and converted to datetime format. The year of each check-in is also extracted and added as a new column. The 'df_checkins_id' dataframe is then grouped by the business ID and year, and the count of check-ins for each group is calculated. Finally, the 'df_checkins_id' dataframe is appended to the 'df_checkins' dataframe.

The code above is creating a new DataFrame called df_checkins and filling it with the check-in data for each business.

In [99]:
# Pickle results to avoid running all the above code again - it took hours to finish running
df_checkins.to_pickle('../data/processed/df_checkins.pickle')

In [18]:
# Load data
df_checkins = pd.read_pickle('../data/processed/df_checkins.pickle')
df_checkins.head()

Business_id  Year Checkin_date
                                       count
0  ---kPU91CF4Lq2-WlRu9Lw  2020            5
1  ---kPU91CF4Lq2-WlRu9Lw  2021            6
0  --0iUa4sNDFiZFrAdIWhZQ  2010            1
1  --0iUa4sNDFiZFrAdIWhZQ  2011            2
2  --0iUa4sNDFiZFrAdIWhZQ  2012            1

This code is creating a new dataframe with information on the number of check-ins per business and year.

In [19]:
# Check the DataFrame shape
df_checkins.shape

(9078, 3)

The DataFrame has 9078 rows and 3 columns.

In [20]:
# Rename the column to make it more clear
df_checkins.rename(columns={'number_checkins': 'checkins_peryear'}, inplace=True)

In [21]:
# Current DataFrame structure: one check-in information per row
# Group the checkin information by business_id
total_checkins = pd.DataFrame(df_checkins.groupby('Business_id')['Year'].sum()).reset_index()

In [22]:
# Estimate the number of years the business is listed on Yelp
total_checkins_yearcount = pd.DataFrame(df_checkins.groupby('Business_id')['Year'].count()).reset_index()

In [23]:
# Put all data in a DataFrame
df_checkins_ver2 = total_checkins.merge(total_checkins_yearcount, on='Business_id')
df_checkins_ver2.columns=['business_id', 'total_checkins', 'total_checkinyear_count']
df_checkins_ver2.head()

business_id  total_checkins  total_checkinyear_count
0  ---kPU91CF4Lq2-WlRu9Lw            4041                        2
1  --0iUa4sNDFiZFrAdIWhZQ           10060                        5
2  --30_8IhuyMHbSOcNWd6DQ            4027                        2
3  --7PUidqRWpRSpXebiyxTg           10065                        5
4  --7jw19RH9JKXgFohspgQw           10083                        5

In [24]:
# Put all data in a DataFrame
restaurants_ver3 = df_checkins_ver2.merge(restaurants_ver2, on='business_id')

The code above is creating a new DataFrame with information about the total number of check-ins and the number of years a business has been listed on Yelp for each business ID. It starts by renaming a column in the df_checkins DataFrame to make it more clear. Then, it calculates the total number of check-ins for each business ID by grouping the df_checkins DataFrame by the 'business_id' column and summing the 'checkins_peryear' column. It also calculates the number of years a business has been listed on Yelp by grouping the df_checkins DataFrame by the 'business_id' column and counting the 'year' column. Finally, it merges these two new DataFrames, which contain the total number of check-ins and the number of years a business has been listed on Yelp, with the restaurants_ver2 DataFrame on the 'business_id' column, creating a new DataFrame called restaurants_ver3.

## Create a column for estimated operating years

In [25]:
# Create a copy to avoid overwrting
df_operatingyears = checkins.copy()

In [26]:
# Check how the DataFrame looks like
df_operatingyears[['business_id', 'date']].head()

business_id                                               date
0  ---kPU91CF4Lq2-WlRu9Lw  2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...
1  --0iUa4sNDFiZFrAdIWhZQ  2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...
2  --30_8IhuyMHbSOcNWd6DQ           2013-06-14 23:29:17, 2014-08-13 23:20:22
3  --7PUidqRWpRSpXebiyxTg  2011-02-15 17:12:00, 2011-07-28 02:46:10, 2012...
4  --7jw19RH9JKXgFohspgQw  2014-04-21 20:42:11, 2014-04-28 21:04:46, 2014...

In [27]:
# Find the first review year

minyear = []

def minyear_extract(x):
    data = x.split(',')
    data = [int(x.strip()[:4]) for x in data]
    minyear.append(min(data))
    return min(data)

In [28]:
# Create a column that contains first review year information
df_operatingyears['minyear'] = df_operatingyears.date.apply(minyear_extract)

In [29]:
# Find the latest review year

maxyear = []

def maxyear_extract(x):
    data = x.split(',')
    data = [int(x.strip()[:4]) for x in data]
    maxyear.append(max(data))
    return max(data)

In [30]:
# Create a column that contains latest review year information
df_operatingyears['maxyear'] = df_operatingyears.date.apply(maxyear_extract)

In [31]:
# Estimate the operating years 
df_operatingyears['estimated_opearating_years'] = df_operatingyears['maxyear'] - df_operatingyears['minyear']

In [32]:
# Put all data in a DataFrame
restaurants_ver4 = df_operatingyears.merge(restaurants_ver3, on='business_id')
restaurants_ver4

business_id  \
0    ---kPU91CF4Lq2-WlRu9Lw   
1    --0iUa4sNDFiZFrAdIWhZQ   
2    --7PUidqRWpRSpXebiyxTg   
3    --8IbOsAAxjKRoYsBFL-PA   
4    --MbOh2O1pATkXa7xbU6LA   
..                      ...   
774  -kVGt2QRNtBoZMvKeIjVug   
775  -kWKHxA5tcrCfbfx-_OLhQ   
776  -kYESienondGaj9TFadkUg   
777  -kY_HDP7IMvGl-kBIZVU4A   
778  -ka45SR1nuoxUU4eTamE_w   

                                                  date  minyear  maxyear  \
0    2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...     2020     2021   
1    2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...     2010     2014   
2    2011-02-15 17:12:00, 2011-07-28 02:46:10, 2012...     2011     2015   
3    2015-06-06 01:03:19, 2015-07-29 16:50:58, 2015...     2015     2018   
4    2013-04-21 01:52:06, 2013-05-02 04:02:07, 2013...     2013     2017   
..                                                 ...      ...      ...   
774  2012-03-27 22:27:40, 2012-07-21 00:52:58, 2012...     2012     2014   
775  2011-02-21 18:55:49, 2011-02-27 00:08:03, 2011...     2011     2021   
776  2010-12-10 17:45:46, 2011-01-17 01:49:42, 2011...     2010     2017   
777  2012-10-21 18:42:43, 2012-10-22 00:50:42, 2012...     2012     2022   
778  2016-05-09 13:06:09, 2016-05-27 11:45:10, 2016...     2016     2019   

     estimated_opearating_years  total_checkins  total_checkinyear_count  \
0                             1            4041                        2   
1                             4           10060                        5   
2                             4           10065                        5   
3                             3            8066                        4   
4                             4           10075                        5   
..                          ...             ...                      ...   
774                           2            6039                        3   
775                          10           22176                       11   
776                           7           16108                        8   
777                          10           22187                       11   
778                           3            8070                        4   

                      business_name             city US_state  ...  \
0                 Frankie's Raw Bar  New Port Richey       FL  ...   
1      Pupuseria Y Restaurant Melba        Clementon       NJ  ...   
2        Humpty's Family Restaurant         Edmonton       AB  ...   
3          The Original Italian Pie         Gentilly       LA  ...   
4                      Sweet Lizard           Tucson       AZ  ...   
..                              ...              ...      ...  ...   
774                     TGI Fridays             Avon       IN  ...   
775         Field House Pub & Grill      Saint Louis       MO  ...   
776          Ethio Cafe & Carry Out     Philadelphia       PA  ...   
777  Dune Coffee Roasters - Anacapa    Santa Barbara       CA  ...   
778         Abington All Star Bagel         Abington       PA  ...   

     food_type_Sandwiches  food_type_Seafood  food_type_Soup  \
0                       0                  1               0   
1                       0                  0               0   
2                       0                  0               0   
3                       0                  0               0   
4                       0                  0               0   
..                    ...                ...             ...   
774                     0                  0               0   
775                     1                  0               0   
776                     0                  0               0   
777                     0                  0               0   
778                     1                  0               0   

     food_type_Steakhouses  food_type_Sushi Bars  food_type_Tacos  \
0                        0                     0                0   
1                        0                     0                0   


In [33]:
# Drop the date column since I don't need to use it again
restaurants_ver4.drop(columns='date', inplace=True)

This code creates a new DataFrame df_operatingyears that contains information about the first and last years in which a business received a review, and estimates the number of years that the business has been operating based on these values. It does this by using the minyear_extract and maxyear_extract functions to extract the minimum and maximum review years, respectively, and then subtracting the minimum review year from the maximum review year to find the number of operating years. Finally, it merges this information with the restaurants_ver3 DataFrame and drops the date column.

## Create a column for total review count

In [34]:
# Load review data
reviewdf = pd.read_pickle('../data/processed/reviewdf.pickle')
reviewdf.head(2)

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q  7ATYjTIgM3jUlt4UM3IypQ   

   stars  useful  funny  cool  \
0    3.0       0      0     0   
1    5.0       1      0     1   

                                                text                 date  
0  If you decide to eat here, just be aware it is...  2018-07-07 22:09:11  
1  I've taken a lot of spin classes over the year...  2012-01-03 15:28:18

In [119]:
# Group reviews by business_id
total_reviews = pd.DataFrame(reviewdf.groupby('business_id').count()['review_id'])
total_reviews = total_reviews.reset_index()

In [120]:
# Put all data in a DataFrame
restaurants_ver5 = total_reviews.merge(restaurants_ver4, on='business_id')

This code creates a new dataframe called "total_reviews" which groups the review data by the business_id and counts the number of review_id's for each business. It then resets the index of this dataframe and combines it with the restaurants_ver2 dataframe using the business_id as a common key. The resulting dataframe is called "restaurants_ver5".

# Create predictor variables

In [121]:
# Create categories per business type
# Businesses that is active and run for 3+ years: Successful
# Businesses that is active but run for less than 3 years: Active for less than three years
# Busienss that is closed: Failed
predictions = []

for o in range(len(restaurants_ver5)):
    if (restaurants_ver5['estimated_opearating_years'][o] >= 3) & (restaurants_ver5['business_status'][o] == 1):
        predictions.append('successful')
    elif (restaurants_ver5['estimated_opearating_years'][o] < 3) & (restaurants_ver5['business_status'][o] == 1):
        predictions.append('active for less than three years')
    else:
        predictions.append('failed')

In [122]:
# Create a new column
restaurants_ver5['prediction_col'] = predictions

In [123]:
# Check the distribution
total = len(restaurants_ver5['prediction_col'])
successful_count = len(restaurants_ver5[restaurants_ver5['prediction_col'] == 'successful'])
active_count = len(restaurants_ver5[restaurants_ver5['prediction_col'] == 'active for less than three years'])
failed_count = len(restaurants_ver5[restaurants_ver5['prediction_col'] == 'failed'])

In [124]:
# Remove data that are not necessary for this capstone
# This capstone predicts 'successful' or 'failed', thus other information need to be removed
restaurants_clean = restaurants_ver5[restaurants_ver5['prediction_col'] != 'active for less than three years']

In [131]:
p_restaurantss_clean.drop(columns=['address', 'hours', 'postal_code'], inplace=True)
p_restaurantss_clean


business_id  review_id  minyear  maxyear  \
1      --0iUa4sNDFiZFrAdIWhZQ         14     2010     2014   
2      --7PUidqRWpRSpXebiyxTg         12     2011     2015   
3      --8IbOsAAxjKRoYsBFL-PA         27     2015     2018   
4      --MbOh2O1pATkXa7xbU6LA         27     2013     2017   
5      --SJXpAa0E-GCp2smaHf0A         13     2014     2019   
...                       ...        ...      ...      ...   
63301  zzjCxn89a7RQo8keIOO_Ag          5     2011     2015   
63302  zzjFdJwXuxBOGe9JeY_EMw         48     2015     2021   
63303  zznJox6-nmXlGYNWgTDwQQ         30     2013     2021   
63305  zzu6_r3DxBJuXcjnOYVdTw          8     2011     2013   
63306  zzyx5x0Z7xXWWvWnZFuxlQ          8     2015     2015   

       estimated_opearating_years  total_checkins  total_checkinyear_count  \
1                               4           10060                        5   
2                               4           10065                        5   
3                               3            8066                        4   
4                               4           10075                        5   
5                               5           12099                        6   
...                           ...             ...                      ...   
63301                           4            6038                        3   
63302                           6           14126                        7   
63303                           8           14115                        7   
63305                           2            6036                        3   
63306                           0            2015                        1   

                      business_name          city US_state  ...  \
1      Pupuseria Y Restaurant Melba     Clementon       NJ  ...   
2        Humpty's Family Restaurant      Edmonton       AB  ...   
3          The Original Italian Pie      Gentilly       LA  ...   
4                      Sweet Lizard        Tucson       AZ  ...   
5                        Winn Dixie     Riverview       FL  ...   
...                             ...           ...      ...  ...   
63301                     Thriftway     Pottstown       PA  ...   
63302          Taste of the Islands    Norristown       PA  ...   
63303                       Dunkin'    Clearwater       FL  ...   
63305                  Cafe Diblasi        Gretna       LA  ...   
63306           Walnut Street Pizza  Philadelphia       PA  ...   

       food_type_Seafood  food_type_Soup  food_type_Steakhouses  \
1                      0               0                      0   
2                      0               0                      0   
3                      0               0                      0   
4                      0               0                      0   
5                      0               0                      0   
...                  ...             ...                    ...   
63301                  0               0                      0   
63302                  0               0                      0   
63303                  0               0                      0   
63305                  0               0                      0   
63306                  0               0                      0   

       food_type_Sushi Bars  food_type_Tacos  food_type_Tapas Bars  \
1                         0                0                     0   
2                         0                0                     0   
3                         0                0                     0   
4                         0                0                     0   
5                         0                0                     0   
...                     ...              ...                   ...   
63301                     0                0                     0   
63302                     0                0                     0   
63303                     0                0                     0   
63305                     

In [132]:
# Pickle results to avoid running all cells 
p_restaurantss_clean.to_pickle('../data/processed/restaurants_clean.pickle')

In [129]:
# Load data
p_restaurants_clean = pd.read_pickle('../data/processed/restaurants_cleann.pickle')
p_restaurants_clean

business_id  review_id  minyear  maxyear  \
1    --0iUa4sNDFiZFrAdIWhZQ         14     2010     2014   
2    --7PUidqRWpRSpXebiyxTg         12     2011     2015   
3    --8IbOsAAxjKRoYsBFL-PA         27     2015     2018   
4    --MbOh2O1pATkXa7xbU6LA         27     2013     2017   
5    --SJXpAa0E-GCp2smaHf0A         13     2014     2019   
..                      ...        ...      ...      ...   
774  -kVGt2QRNtBoZMvKeIjVug          7     2012     2014   
775  -kWKHxA5tcrCfbfx-_OLhQ         89     2011     2021   
776  -kYESienondGaj9TFadkUg        116     2010     2017   
777  -kY_HDP7IMvGl-kBIZVU4A        330     2012     2022   
778  -ka45SR1nuoxUU4eTamE_w         58     2016     2019   

     estimated_opearating_years  total_checkins  total_checkinyear_count  \
1                             4           10060                        5   
2                             4           10065                        5   
3                             3            8066                        4   
4                             4           10075                        5   
5                             5           12099                        6   
..                          ...             ...                      ...   
774                           2            6039                        3   
775                          10           22176                       11   
776                           7           16108                        8   
777                          10           22187                       11   
778                           3            8070                        4   

                      business_name           city US_state  ...  \
1      Pupuseria Y Restaurant Melba      Clementon       NJ  ...   
2        Humpty's Family Restaurant       Edmonton       AB  ...   
3          The Original Italian Pie       Gentilly       LA  ...   
4                      Sweet Lizard         Tucson       AZ  ...   
5                        Winn Dixie      Riverview       FL  ...   
..                              ...            ...      ...  ...   
774                     TGI Fridays           Avon       IN  ...   
775         Field House Pub & Grill    Saint Louis       MO  ...   
776          Ethio Cafe & Carry Out   Philadelphia       PA  ...   
777  Dune Coffee Roasters - Anacapa  Santa Barbara       CA  ...   
778         Abington All Star Bagel       Abington       PA  ...   

     food_type_Seafood  food_type_Soup  food_type_Steakhouses  \
1                    0               0                      0   
2                    0               0                      0   
3                    0               0                      0   
4                    0               0                      0   
5                    0               0                      0   
..                 ...             ...                    ...   
774                  0               0                      0   
775                  0               0                      0   
776                  0               0                      0   
777                  0               0                      0   
778                  0               0                      0   

     food_type_Sushi Bars  food_type_Tacos  food_type_Tapas Bars  \
1                       0                0                     0   
2                       0                0                     0   
3                       0                0                     0   
4                       0                0                     0   
5                       0                0                     0   
..                    ...              ...                   ...   
774                     0                0                     0   
775                     0                0                     0   
776                     0                0                     0   
777                     0                0                     0   
778                     0           

The code creates a new column in the DataFrame called "prediction_col" that categorizes each restaurant as 'successful', 'active for less than three years', or 'failed'. The categorization is based on the estimated operating years of the restaurant and its business status. If the restaurant has been operating for at least 3 years and has a business status of 1, it is classified as 'successful'. If it has been operating for less than 3 years and has a business status of 1, it is classified as 'active for less than three years'. Otherwise, it is classified as 'failed'. The code then removes the data for restaurants classified as 'active for less than three years' and saves the resulting DataFrame as a pickle file.